### DISCUSSION TITLE
How do I pass a string from VBA to a tB standard dll?

### DISCUSSION THREAD

**Discussion by Greedquest** [twinbasic/twinbasic#1448] [Q&A] [answered]
Maybe being stupid here. I've got a function in tB

```vba
[ DllExport ]
Public Function GetFullAccessor(ByVal moduleName As String, ByVal proj As VBProject, ByRef outPrivateTI As ITypeInfo) As Object
```

I called it from VBA (64-bit) like this:
```vba
'option 1
Public Declare PtrSafe Function GetFullAccessor Lib "C:\GitHub\vbInvoke\Build\vbInvoke_win64.dll" ( _
    ByVal moduleName As String, ByVal proj As VBProject, ByRef outPrivateTI As IUnknown _
) As Object

Sub t()
    Dim accessor As Object
    Dim outTi As IUnknown
    Set accessor = GetFullAccessor("ExampleModule", ThisWorkbook.VBProject, outTi)
End Sub
```

But that crashes excel by sending a random length string of unicode (japanese) characters

Then I tried
```vba
'option 2
Public Declare PtrSafe Function GetFullAccessor Lib "C:\GitHub\vbInvoke\Build\vbInvoke_win64.dll" ( _
    ByVal moduleName As LongPtr, ByVal proj As VBProject, ByRef outPrivateTI As IUnknown _
) As Object

'[...]
Set accessor = GetFullAccessor(StrPtr("ExampleModule"), ThisWorkbook.VBProject, outTi)
```

Which works. Reading [this on SO](https://stackoverflow.com/a/43423527/6609896) I think what's happening is that in option 1, the string is being converted to ANSI (narrow chars) in the declare, but tB expects unicode and so messes up the interpretation resulting in all the garbled characters. Passing pointers in option 2 just works but I don't want my client VBA to have to use pointers.

**Can I declare my tB function in such a way that the VBA declare can pass a string not a `StrPtr`?** Ideally the tB code would also be able to work with Strings, but I'm more keen to have the API be simple than the internal implementation.

**Comment by WaynePhillipsEA** [comment, marked as answer]
Yes, the ANSI conversion is the problem, at the VBA side.   Declare the parameters as `Variant` is usually the easiest solution to avoid the problems.
